### ISSUE TITLE
Missing Accessible Feedback on Page Change 

### ISSUE THREAD

**Issue by johnhbenetech** [benetech/MathShare#742]
When browsing websites which render their content page by page on the server, clicking a link usually causes an entirely new web document to be loaded by the browser.  Screen readers are able to detect this event, inform users of the newly loaded page, and place their keyboard focus at a sensible point. 
 
Because MathShare is implemented as a single page app (SPA), none of this behavior occurs automatically and no code has been implemented to re-create it.  This has a serious impact on the usability of the application for screen reader users. 
 
When a link or button is clicked to load a new content view, there is no speech feedback to indicate that something has happened.  Keyboard focus is not managed at all, meaning that the browser and screen reader will attempt to maintain the user’s current position.  As the page content may have completely changed, this makes the behavior unpredictable.  These problems also occur while users are attempting to navigate back through their history.

#### Suggested Solution 
 
When a user navigates to a route, the application should send the title of the new page to be announced by the screen reader.  Currently, this is not possible because all routes inside MathShare present the same page title.  Please consult issue #743 and then use the titles determined while following the advice there for speech announcements. 
 
In addition, the user’s keyboard focus must be managed when new content is loaded.  As a minimum stop-gap measure, it can be placed on the first element of the page, which re-creates the behavior exhibited by most screen readers when a new page is loaded from the server.  Ultimately, however, every content view in MathShare should have a logical entry point (which should usually be a level 1 heading), and keyboard focus should be placed there instead. 
 
Going further, many screen readers will attempt to remember a user’s position when they return to a previously visited page.  there are opportunities for MathShare to implement this in some situations, such as placing focus on the most recently accessed problem when returning to the list of problems in a set. 
 
It’s worth us reiterating the advice from the “Missing Skip Link” section of this document: keyboard focus should never be placed on an entire page section or container (such as a `<main>` or `<div>`), as a screen reader will attempt to voice all of the child content once it receives focus.  This will be annoying or confusing at best, but can also degrade performance as the software attempts to fill its speech queue and audio synthesis buffer with a large amount of data.

**Comment by alexrcabral** [comment]
@jscholes Could you check over the implementation of this?

**Comment by jscholes** [comment]
There are commits such as https://github.com/benetech/MathShare/commit/3b315e2cfaa7e0a7bd95df059ead6dc75ec6e1d1 and https://github.com/benetech/MathShare/commit/78108c95bbe4a3a9814a0e14ccaa3f25e3477760, but those seem to relate to #743 rather than this ticket.

Adding unique page titles is virtually a requirement before implementing spoken feedback of a change in context.  But the presence of page titles will not result in automatic feedback on its own.

@rupeshparab - are there commits I'm missing, in relation to this issue specifically, which prompted its move into QA?

**Comment by jscholes** [comment]
@johnhbenetech If the main heading receives focus and is descriptive, I think it would be okay for the live region to be "polite".  It would still be there if, for whatever reason, the focus management failed or there was no heading, but it wouldn't interrupt the reading of the heading otherwise.

**Comment by rupeshparab** [comment]
@jscholes 
> @rupeshparab - are there commits I'm missing, in relation to this issue specifically, which prompted its move into QA?

I tried searching for it, I wasn't able to find any commits related to pushing the title to aria-live.
I think I mistook this issue for something else while moving it to QA

Also, I am thinking whether announcing title manually can cause any conflict with the auto-focused heading being read.

**Comment by jscholes** [comment]
@rupeshparab
> I am thinking whether announcing title manually can cause any conflict with the auto-focused heading being read.

It shouldn't cause a conflict.  Feel free to put responsibility on the screen reader to work this out for itself.  But I do think using a "polite" region instead of "assertive" will reduce the likelihood of any issues.

As with anything, problems may become evident once we have an initial implementation.

**Comment by rupeshparab** [comment]
@jscholes 
I have added announcement using aria-live, it seems to be working fine on Voice Over.
Although, for some cases aria-live and focused element have similar text being read.
An example is when the problem set is opened, focused element value is problem set name - 'Solve for X' and page title is 'Solve for X - Benetech Mathshare'
Focused element is read first and then the aria-live text is being read.

**Comment by jscholes** [comment]
@rupeshparab - having some crossover between the heading text and speech announcement is fine, because they act as fallbacks for each other.  That is, if a screen reader fails to announce the focus movement, there's still a speech announcement.  And if a screen reader ignores the announcement, there's still the focus change.  So for the most part, this is working great.  However:

* The "- Benetech Mathshare" suffix shouldn't be included in that announcement.  It should only include the information before it.
* Could we move the live region to the bottom of the DOM, where it's less likely to actually be encountered by a screen reader user?  Ideally we would hide it altogether, but that unfortunately doesn't work.  So all we can do is minimise the chances that someone will come across it.
* When I activate a problem set, the heading which receives focus seems to have the text, "Benetech's MathShare Editor (alpha)".  But the text then updates to the actual name of the set.  Any idea what's going on there?
* It would be great if, when returning to a list of data, the most recently visited item received focus instead of the same heading each time.  For instance, the most recently visited set when pressing back to go to the dashboard, and likewise for returning to the list of problems in a set.  Note that if this is implemented, it should always be a link or button which receives focus, rather than a list item or container.

**Comment by rupeshparab** [comment]
> But the text then updates to the actual name of the set. Any idea what's going on there?

The reason for this was that before the set was loaded a default text was being shown in the title's, I have changed it such that it won't be shown until the api call is complete

The first 3 issues have been fixed, I am trying to build a solution to maintain focus during page transition

**Comment by jscholes** [comment]
@rupeshparab - understood.  This is working well, so I'll close this issue and open a new one related to stateful accessible transitions.

**Comment by jscholes** [comment]
Closed in https://github.com/benetech/MathShare/commit/e224a90b818b420ab648b90be911df51d35f5275.

**Comment by jscholes** [comment]
@rupeshparab - there seems to be some regressions relating to this behaviour.

1. When I activate a link, the live region communicates the title of the destination straight away.  But the level 1 heading doesn't receive focus for around 4-5 seconds.
2. When I activate the "Partners", "Privacy Policy" or "Open Source" links in the footer, the logo receives focus rather than the `<h1>` on the destination page.
3. When I activate the link for an individual problem, return to the problem set page and then choose another problem, the number of the first problem I clicked on is incorrectly spoken, followed by the correct number of the one I just activated.

To reproduce the last of these:

1. Go to the "Distance Rate and Time" set.
2. Activate the first problem.
3. Go back to the set.
4. Activate the second problem.

After step 4, you should hear: "Problem 2 of 3".  But instead, you will hear, "Problem 1 of 3, problem 2 of 3".

**Comment by jscholes** [comment]
@rupeshparab - a couple of other observations:

1. If focus has been placed on the heading of an alert, e.g. the "Success" message which appears after sign in, the main Dashboard heading shouldn't then receive focus later.  This is confusing and won't give users enough time to read the alert.
2. When activating the "sign in" and "sign up" links, no heading receives focus.

**Comment by rupeshparab** [comment]
I have added a check to not move focus to other elements while an alert is in focus. For the second issue, I think it must have got fixed in some other ticket, currently Sign In and Sign Up headings are being focused.

**Comment by johnhbenetech** [comment]
@bpmcneilly please review changes in QA when you get a chance

**Comment by jscholes** [comment]
@rupeshparab - Do you have any thoughts relating to:

> 1. When I activate a link, the live region communicates the title of the destination straight away. But the level 1 heading doesn't receive focus for around 4-5 seconds.

... and ...

> 3. When I activate the link for an individual problem, return to the problem set page and then choose another problem, the number of the first problem I clicked on is incorrectly spoken, followed by the correct number of the one I just activated.

The delay in particular is a severe accessibility issue, as it also ends up interfering with a user's activity if they've already started reading the page content before focus is placed.  The content seems to be ready much earlier than the focus movement occurs.

**Comment by jscholes** [comment]
@rupeshparab Any chance we can move this up on the roadmap?  It is a critical issue for all screen reader users, whether students or staff, and could also have an impact on lots of other users too.  It is confusing for focus to move after such a lengthy delay, when many users may have already started navigating or interacting with the page content.

**Comment by rupeshparab** [comment]
I tried reducing the wait time, on voiceover it causes the title to be interrupted midway and the level 1 heading starts. Will check if there is any way to prevent the live region speech be overridden when the focus moves.
Will also try both of the issues on NVDA

**Comment by jscholes** [comment]
@rupeshparab NVDA won't interrupt itself like VoiceOver.  But what if you do the focus movement first?

**Comment by rupeshparab** [comment]
@jscholes 
Right now the focus movement is on basis of trial and error
There is a retrial mechanism which tries to find a focusable element and if it is not able to find it (in cases where view has not rendered), it will sleep for a few ms before retrying. This sleep time increases after every retrial. I can reduce the amount of sleep, but it cannot guarantee that this focus will happen before the title gets announced in aria-live

**Comment by jscholes** [comment]
@rupeshparab That's fine; we don't really need to work around a VoiceOver-specific thing.  The critical issue to resolve is the long delay between the two announcements, so that the focus movement doesn't happen after several seconds.

Having said that, I don't understand why such a long delay is ever being reached.  In all of my testing, the heading to receive focus has been present in the DOM long before the focus movement actually occurs, so it seems like there may be some errors in the exponential backoff logic.

**Comment by rupeshparab** [comment]
right, I too think the retrials might be increasing the sleep too much
will set a max beyond which the sleep time should remain constant

will check if that decreases the delay

**Comment by rupeshparab** [comment]
@jscholes 
I have pushed few fixes on Mathshare QA to improve focus
Can you check if you are still facing issues related to delayed focus?

**Comment by jscholes** [comment]
@rupeshparab Seems to be fixed.  Will reopen if I come across any further issues during testing.
